code/mission/missioncampaign.cpp

```diff
@@ -1863,6 +1918,42 @@ void mission_campaign_save_on_close_variables()
 	mission_campaign_store_variables(SEXP_VARIABLE_SAVE_ON_MISSION_CLOSE, false);
 }
 
+// jg18 - adapted from mission_campaign_save_on_close_variables()
+void mission_campaign_save_on_close_containers()
+{
+	// make sure we are actually playing a single-player campaign
+	if (!(Game_mode & GM_CAMPAIGN_MODE) || (Campaign.type != CAMPAIGN_TYPE_SINGLE))
+		return;
+
+	// now save containers
+	for (const auto &container : get_all_sexp_containers()) {
+		// we only want the on mission close type. On campaign progress type are dealt with elsewhere
+		if (none(container.type & ContainerType::SAVE_ON_MISSION_CLOSE)) {
+			continue;
+		}
+
+		// deal with eternals
+		if (container.is_eternal()) {
+			// check if container already exists and update it
+			auto ppc_it = std::find_if(Player->containers.begin(),
+				Player->containers.end(),
+				[container](const sexp_container& ppc) {
+					return !stricmp(container.container_name.c_str(), ppc.container_name.c_str());
```

Comment:
Same question a third time, about whether container comparison here should take container type into account and not just name.